engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala

```diff
@@ -6,37 +6,117 @@ import scala.annotation.compileTimeOnly
+  * and IDE tooling cannot help you track these errors down and so you are bound to find out about them at runtime, with
+  * potentially uninformative error messages.
+  *
+  * If instead, you use these macros, these troubles will be taken away. Define your `Fact`s like this and have the IDE and
+  * compiler warn you about duplicate names etc:
+  *
+  * `val MyFact = FactMacros.defineFact[Int]("Description")`
+  *
+  */
 object FactMacros {
 
+  // Macro definitions
+
+  /**
+    * Defines a `Fact`, using the name of the `val` it is assigned to as the name of the `Fact`. Note: the value of this
+    * macro *must* be assigned to a `val`, otherwise a compiler error will be raised.
+    *
+    * @tparam A the value type of the resulting `Fact`.
+    * @return a `SingularFact` initialized with the name of the val declaration.
+    */
+  def defineFact[A](): SingularFact[A] = macro FactMacros.defineFactMacroImpl[A]
+
+  /**
+    * Defines a `Fact`, using the name of the `val` it is assigned to as the name of the `Fact`. Note: the value of this
+    * macro *must* be assigned to a `val`, otherwise a compiler error will be raised.
+    *
+    * @tparam A the value type of the resulting `Fact`.
+    * @param description description of the `Fact`, to be passed along to the `Fact`'s constructor.
+    * @return a `SingularFact` initialized with the name of the val declaration.
+    */
+  def defineFactWithDescription[A](description: String): SingularFact[A] = macro FactMacros.defineFactMacroWithDescriptionImpl[A]
+
+  /**
+    * Defines a `Fact`, using the name of the `val` it is assigned to as the name of the `Fact`. Note: the value of this
+    * macro *must* be assigned to a `val`, otherwise a compiler error will be raised.
+    *
+    * @tparam A the value type of the resulting `Fact`.
+    * @return a `ListFact` initialized with the name of the val declaration.
+    */
+  def defineListFact[A](): ListFact[A] = macro FactMacros.defineListFactMacroImpl[A]
+
+  /**
+    * Defines a `Fact`, using the name of the `val` it is assigned to as the name of the `Fact`. Note: the value of this
+    * macro *must* be assigned to a `val`, otherwise a compiler error will be raised.
+    *
+    * @tparam A the value type of the resulting `Fact`.
+    * @param description description of the `Fact`, to be passed along to the `Fact`'s constructor.
+    * @return a `ListFact` initialized with the name of the val declaration.
+    */
+  def defineListFactWithDescription[A](description: String): ListFact[A] = macro FactMacros.defineListFactMacroWithDescriptionImpl[A]
+
+
+  // Macro implementations
+
   @compileTimeOnly("This is a compile-time macro implementation, do not call it at runtime")
-  def defineFactMacroImpl[A : c.WeakTypeTag](c: Context)(omschrijving: c.Expr[String]): c.Expr[SingularFact[A]] =
+  def defineFactMacroImpl[A : c.WeakTypeTag](c: Context)(): c.Expr[SingularFact[A]] =
```

Comment:
Nope, as far as macros go, this is a bit of a convention. Because a macro consists of a macro *definition* and a macro *implementation*. To keep these apart, `Impl` is added as a postfix in pretty much all of the examples. Note that these `Impl`-methods are not supposed to be used on runtime. (Hence the `@compileTimeOnly` annotations).